### PR TITLE
Add a placeholder for the sns-cache code

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
We would like to add the sns cache to this codebase to keep the cache in sync with the nns-dapp.

# Changes
- Add a Rust placeholder for the nns-dapp
- Bump the Rust version to match the sns-cache code.
- Fix clippy lints exposed by the Rust version bump.
- Use the same rust version for the nns-dapp and the cache. If the two are meant to be kept in sync we can't have breaking changes in one without changes in the other so it probably makes sense to link the versions, just as say the candid and  candid-macro crates have the same version number.

# Future
- Add the cache to dfx.json
- Deploy the cache along with the nns-dapp when deploying to testnets and in CI
- Fill in the cache rust code
- Fill in the ic-repo type derivation from candid

# Tests
- This should compile and not break the nns-dapp rust code.  No more.

### CODE REVIEW

**Review by lmuntaner — APPROVED** [review]
Thanks!

The reproducibility builds pass in this PR!

File: rs/backend/src/ledger_sync.rs [hunk 1]
```diff
@@ -51,12 +51,7 @@ async fn sync_transactions_within_lock() -> Result<u32, String> {
             let blocks_count = blocks.len() as u32;
             for (block_height, block) in blocks.into_iter() {
                 let transaction = block.transaction().into_owned();
-                let result =
```
Comment (peterpeterparker) [inline rs/backend/src/ledger_sync.rs:54]: why this change?

**Review by peterpeterparker — COMMENTED** [review]
Shouldn't the sns_cache be declared in dfx.json?

e.g. nns-dapp

```
"nns-dapp": {
      "type": "custom",
      "candid": "rs/nns-dapp.did",
      "wasm": "nns-dapp.wasm",
      "build": "./build.sh"
    },
```

File: Dockerfile [hunk 2]
```diff
@@ -54,7 +54,8 @@ RUN cargo install --version 0.3.1 ic-cdk-optimizer
 COPY Cargo.lock .
 COPY Cargo.toml .
 COPY rs/backend/Cargo.toml rs/backend/Cargo.toml
-RUN mkdir -p rs/backend/src && touch rs/backend/src/lib.rs && cargo build --target wasm32-unknown-unknown --release --package nns-dapp && rm -rf rs/backend/src
+COPY rs/sns_cache/Cargo.toml rs/sns_cache/Cargo.toml
+RUN mkdir -p rs/backend/src rs/sns_cache/src && touch rs/backend/src/lib.rs && touch rs/sns_cache/src/lib.rs && cargo build --target wasm32-unknown-unknown --release --package nns-dapp && rm -rf rs/backend/src
```
Comment (peterpeterparker) [inline Dockerfile:58]: is `rm -rf rs/sns_cache/src` missing?

File: rs/backend/src/ledger_sync.rs (hunk 1 above)
Comment (peterpeterparker) [reply rs/backend/src/ledger_sync.rs:54]: Nvm missed the `? `

File: Dockerfile (hunk 2 above)
Comment (bitdivine) [reply Dockerfile:58]: Good catch.  Adding...

**Comment by bitdivine** [comment]
> Shouldn't the sns_cache be declared in dfx.json?
@peterpeterparker  Not in this PR; please see the PR description:
> # Future
> * Add the cache to dfx.json

**Comment by bitdivine** [comment]
> The reproducibility builds pass in this PR!

@lmuntaner That is thanks to @dskloetd 's PR being merged.
